Dockerfile

```diff
@@ -1,16 +1,28 @@
-FROM debian:bookworm-slim
+FROM python:3.11-slim-trixie AS joinmarket
 
-RUN mkdir -p /jm/clientserver
 WORKDIR /jm/clientserver
+COPY ./pubkeys ./pubkeys
+COPY ./install.sh ./install.sh
+
+RUN apt-get update \
+  && apt-get install -y --no-install-recommends \
+    gnupg \
+    ca-certificates \
+    curl \
+  && ./install.sh --docker-install --deps-only \
+  && apt-get clean \
+  && rm -rf /var/lib/apt/lists/*
 
 COPY . .
 
-RUN apt-get update && apt-get install -y --no-install-recommends gnupg ca-certificates=* curl=* \
-  python3-pip=* python3=* \
-  && pip3 config set global.break-system-packages true \
-  && pip3 install 'wheel>=0.35.1' \
-  && ./install.sh --docker-install \
-  && apt-get purge -y --autoremove python3-pip \
+RUN apt-get update \
+  && ./install.sh --docker-install --no-deps \
   && apt-get clean \
   && rm -rf /var/lib/apt/lists/*
 
+FROM bitcoin/bitcoin:29.2 AS bitcoin
+FROM joinmarket AS test
+COPY --from=bitcoin /opt/bitcoin-29.2/bin /usr/local/bin/
+RUN ./install.sh --docker-install --no-deps --develop
+
+FROM joinmarket AS final
```

Comment:
IMO a major benefit of multi-stage builds is to do the build in a separate branch from the final image, basing the final image on a clean base such as `debian:slim` or `python:3.12-slim`, and copy over the necessary artefacts from the build image (like you are doing with `bitcoin` here). Since build scripts have never been run in the context of the final image this way, this can have both security and space-efficiency benefits.

Did you consider this approach? What's the benefit of this approach of multi-stage where the final image is simply a copy of the build step?

Reply:
For now, this multi layer approach allows to extend the main layer with test dependencies, no more. Ideally we indeed want a thin final layer, which requires a refactoring of the install script imo, in a subsequent PR.